### PR TITLE
Converting SDLNames Defines to Constants

### LINKED ISSUE #3: Use Constants instead of #define
The parameter key and value constants defined in SDLNames.h should use extern instead of define for safe compile time type checking and to avoid collisions.

[Apple's Recommendation](https://developer.apple.com/library/mac/documentation/cocoa/conceptual/codingguidelines/Articles/NamingIvarsAndTypes.html#//apple_ref/doc/uid/20001284-1003095) suggests

>Define constants for strings used for such purposes as notification names and dictionary keys. By using string constants, you are ensuring that the compiler verifies the proper value is specified (that is, it performs spell checking).

### PR DESCRIPTION
**Pull request by asm09fsu**
Fixes #3 

This PR is **ready** for review.

### Risk
This PR makes **no** API changes.

### Testing Plan
These changes have 100% passing tests.

### Summary
This PR changes all `#define`s in SDLNames.h to use a typedef of `NSString*` called `SDLName`. Also, `SDLFunctionID` now uses a static `NSDictionary` for function names/ids to improve performance, and only one should ever need to be allocated.

##### Enchancements
* Updated all #defines in SDLNames.h to use constants.
* Updated SDLFunctionID to use a static dictionary.

### CODE REVIEW

File: SmartDeviceLink-iOS.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -505,7 +505,7 @@
 		5D61FCF91A84238C00846EE7 /* SDLMenuParams.m in Sources */ = {isa = PBXBuildFile; fileRef = 5D61FB0C1A84238A00846EE7 /* SDLMenuParams.m */; };
 		5D61FCFA1A84238C00846EE7 /* SDLMyKey.h in Headers */ = {isa = PBXBuildFile; fileRef = 5D61FB0D1A84238A00846EE7 /* SDLMyKey.h */; settings = {ATTRIBUTES = (Public, ); }; };
 		5D61FCFB1A84238C00846EE7 /* SDLMyKey.m in Sources */ = {isa = PBXBuildFile; fileRef = 5D61FB0E1A84238A00846EE7 /* SDLMyKey.m */; };
-		5D61FCFC1A84238C00846EE7 /* SDLNames.h in Headers */ = {isa = PBXBuildFile; fileRef = 5D61FB0F1A84238A00846EE7 /* SDLNames.h */; };
+		5D61FCFC1A84238C00846EE7 /* SDLNames.h in Headers */ = {isa = PBXBuildFile; fileRef = 5D61FB0F1A84238A00846EE7 /* SDLNames.h */; settings = {ATTRIBUTES = (Public, ); }; };
```
Comment (joeljfischer) [inline SmartDeviceLink-iOS.xcodeproj/project.pbxproj]: Does this have to be public now?

File: SmartDeviceLink/SDLFunctionID.m [hunk 2]
```diff
@@ -4,36 +4,93 @@
...
 #import "SDLFunctionID.h"
 
-#import "SDLNames.h"
+@interface SDLFunctionID ()
+
+@property (nonatomic, strong, nonnull) NSDictionary* functionIDs;
+
+@end
 
 @implementation SDLFunctionID
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLFunctionID.m]: This class can be private

File: SmartDeviceLink/SDLFunctionID.m [hunk 3]
```diff
@@ -4,36 +4,93 @@
...
+                            };
+        }
+        self.functionIDs = functionIDs;
     }
     return self;
 }
 
-- (NSString *)getFunctionName:(int)functionID {
-    return [functionIDs objectForKey:[NSString stringWithFormat:@"%d", functionID]];
+- (SDLName)getFunctionName:(int)functionID {
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLFunctionID.m]: The `get` function naming is very Java. Should probably be more like `functionNameForId:`

File: SmartDeviceLink/SDLGPSData.m [hunk 4]
```diff
@@ -24,110 +24,110 @@ - (instancetype)initWithDictionary:(NSMutableDictionary *)dict {
...
 
 - (NSNumber *)latitudeDegrees {
-    return [store objectForKey:NAMES_latitudeDegrees];
+    return [store objectForKey:SDLNameLatitudeDegrees];
 }
 
 - (void)setUtcYear:(NSNumber *)utcYear {
     if (utcYear != nil) {
-        [store setObject:utcYear forKey:NAMES_utcYear];
+        [store setObject:utcYear forKey:SDLNameUtcYear];
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLGPSData.m]: Same as above comment regarding capitalization?

**Comment by asm09fsu** [comment]
@joeljfischer I think this is ready, can you validate?

File: SmartDeviceLink-iOS.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -427,7 +427,7 @@
 		5D61FCAA1A84238C00846EE7 /* SDLFileType.m in Sources */ = {isa = PBXBuildFile; fileRef = 5D61FABD1A84238A00846EE7 /* SDLFileType.m */; };
 		5D61FCAB1A84238C00846EE7 /* SDLFuelCutoffStatus.h in Headers */ = {isa = PBXBuildFile; fileRef = 5D61FABE1A84238A00846EE7 /* SDLFuelCutoffStatus.h */; settings = {ATTRIBUTES = (Public, ); }; };
 		5D61FCAC1A84238C00846EE7 /* SDLFuelCutoffStatus.m in Sources */ = {isa = PBXBuildFile; fileRef = 5D61FABF1A84238A00846EE7 /* SDLFuelCutoffStatus.m */; };
-		5D61FCAD1A84238C00846EE7 /* SDLFunctionID.h in Headers */ = {isa = PBXBuildFile; fileRef = 5D61FAC01A84238A00846EE7 /* SDLFunctionID.h */; };
+		5D61FCAD1A84238C00846EE7 /* SDLFunctionID.h in Headers */ = {isa = PBXBuildFile; fileRef = 5D61FAC01A84238A00846EE7 /* SDLFunctionID.h */; settings = {ATTRIBUTES = (Private, ); }; };
```
Comment (joeljfischer) [inline SmartDeviceLink-iOS.xcodeproj/project.pbxproj]: Note, we don't actually want 'private', we want 'project'.

File: SmartDeviceLink/SDLProxy.h [hunk 6]
```diff
@@ -42,7 +42,7 @@ __deprecated_msg("Use SDLManager instead")
 - (void)sendRPCRequest:(SDLRPCMessage *)msg __deprecated_msg("use -sendRPC: instead");
 
 - (void)handleRPCDictionary:(NSDictionary *)dictionary;
-- (void)handleRpcMessage:(NSDictionary *)msg __deprecated_msg("use -handleRPCDictionary: instead");
+- (void)handleRPCMessage:(NSDictionary *)msg __deprecated_msg("use -handleRPCDictionary: instead");
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLProxy.h]: This cannot be changed

File: SmartDeviceLink/SDLProtocol.m [hunk 7]
```diff
@@ -203,7 +203,7 @@ - (BOOL)sendRPC:(SDLRPCMessage *)message encrypted:(BOOL)encryption error:(NSErr
             // Build a binary header
             // Serialize the RPC data into an NSData
             SDLRPCPayload *rpcPayload = [[SDLRPCPayload alloc] init];
-            rpcPayload.functionID = [[[[SDLFunctionID alloc] init] getFunctionID:[message getFunctionName]] intValue];
+            rpcPayload.functionID = [[[SDLFunctionID sharedInstance] functionIdForName:[message functionNameForId]] intValue];
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLProtocol.m]: A few things about this seem odd.

1. Why are we getting the name for the id, then the id for the name? Seems circular?

2. Wouldn't the message's `functionNameForId` method just be `functionName` if we're not passing anything? Can't find if or where that changed in this PR though.

File: SmartDeviceLink/SDLRPCMessage.h [hunk 8]
```diff
@@ -13,7 +13,7 @@
 
 - (instancetype)initWithName:(NSString *)name;
 - (instancetype)initWithDictionary:(NSMutableDictionary *)dict;
-- (NSString *)getFunctionName;
+- (NSString *)functionNameForId;
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLRPCMessage.h]: Oh, here it is, guessing a copy-paste error?

Comment (joeljfischer) [inline SmartDeviceLink/SDLRPCMessage.h]: So to be clear, this has to stay the same to preserve the impact of this PR.

File: SmartDeviceLink/SDLProxy.h (hunk 6 above)
Comment (asm09fsu) [inline SmartDeviceLink/SDLProxy.h]: This was a mis-commit when using regex to modify the 1,000,000 files involved in this PR.

File: SmartDeviceLink/SDLProtocol.m (hunk 7 above)
Comment (asm09fsu) [inline SmartDeviceLink/SDLProtocol.m]: seems we need a way to retrieve the id for a function name (as a response perhaps?) and the function name for the id (to send a request?). This is just a slight modification/improvement over how it's already been done. A full revamp can be done for 5.0.

Comment (joeljfischer) [inline SmartDeviceLink/SDLProtocol.m]: I think i was just confused by the change in `[message functionNameForId]`. It's actually getting the function name not the id.

**Pull request merged**